### PR TITLE
commit for comments

### PR DESCRIPTION
**Pull request by artemalabier**

### CODE REVIEW

File: scraper.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+from analysis import get_exchange_rate
+from data_parser import extract_price
+
+
+class Car:
+    def __init__(self, id, title, link, price):
+        self.id = id
+        self.title = title
+        self.link = link
+        self.price = price
```
Comment (alexpilk) [inline scraper.py:14]: Use a dataclass instead

File: scraper.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+            for car_listing in car_listings:
+                # Extract data for each car listing
+                id = car_listing['id']
+
+                title_elem = car_listing.find('h6', class_='css-16v5mdi er34gjf0')
+                title = title_elem.text.strip() if title_elem else 'N/A'
+
+                link_elem = car_listing.findNext('a', class_='css-rc5s2u')
+                link = link_elem['href'] if link_elem and 'href' in link_elem.attrs else 'N/A'
```
Comment (alexpilk) [inline scraper.py:40]:
```suggestion
                link = link_elem.get('href', 'N/A') if link_elem else 'N/A'
```

File: scraper.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        response = requests.get(url)
+
+        if response.status_code == 200:
+            soup = BeautifulSoup(response.text, 'html.parser')
+
+            # Find all car listings
+            car_listings = soup.find_all('div', {'data-cy': 'l-card'})
+
+            # Create a set to store Car objects
+            cars = set()
```
Comment (alexpilk) [inline scraper.py:30]: The set here doesn't work as you expect it to
```
>>> class Dupa:
...     def __init__(self, a):
...         self.a = a
... 
>>> my_set = set()
>>> my_set.add(Dupa(1))
>>> my_set.add(Dupa(1))
>>> len(my_set)
2
>>> 
```

If you make `Car` a dataclass and set `frozen` & `eq` to True - it will works as expected:
https://docs.python.org/3/library/dataclasses.html

File: scraper.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+            for car_listing in car_listings:
+                # Extract data for each car listing
+                id = car_listing['id']
+
+                title_elem = car_listing.find('h6', class_='css-16v5mdi er34gjf0')
+                title = title_elem.text.strip() if title_elem else 'N/A'
```
Comment (alexpilk) [inline scraper.py:37]: Seems like everything would be easier if you use `None` instead of `'N/A'`

File: main.py [hunk 5]
```diff
@@ -1,16 +1,68 @@
...
-if __name__ == '__main__':
-    print_hi('PyCharm')
+    def collecting_cars(car_urls):
+        all_cars_data = []
+
+        for car_name, urls in car_urls.items():
+            pl_url = urls.get("PL", "")
+            pt_url = urls.get("PT", "")
+
+            pl_data = fetch_olx_car_data(pl_url)
```
Comment (alexpilk) [inline main.py:29]: `fetch_olx_car_data` can sometimes return `None` instead of a set.

Then `list(None)` will throw an error and `list({Car(...), ...})` will convert the set to a list, which also isn't necessary.

I suggest you:
1. Make sure `fetch_olx_car_data` returns a list in all cases
2. Do not convert the results to a list on line 33

File: main.py [hunk 6]
```diff
@@ -1,16 +1,68 @@
...
 
-# Press the green button in the gutter to run the script.
-if __name__ == '__main__':
-    print_hi('PyCharm')
+    def collecting_cars(car_urls):
+        all_cars_data = []
+
+        for car_name, urls in car_urls.items():
+            pl_url = urls.get("PL", "")
+            pt_url = urls.get("PT", "")
```
Comment (alexpilk) [inline main.py:27]: If you extract the code in this loop to a function like `def fetch_cars_for_country(country)` you'll be able to convert the loop to a list comprehension.

File: main.py [hunk 7]
```diff
@@ -1,16 +1,68 @@
-# This is a sample Python script.
+import json
 
-# Press Shift+F10 to execute it or replace it with your code.
-# Press Double Shift to search everywhere for classes, files, tool windows, actions, and settings.
+from analysis import calculate_price_coefficients
+from scraper import fetch_olx_car_data
 
+if __name__ == '__main__':
+    filename = 'URLs.json'
```
Comment (alexpilk) [inline main.py:7]: The idea behind `if __name__ == '__main__':` is to let you import stuff from a module without executing the code in this section.

Therefore it doesn't make sense to define functions in the `if __name__ == '__main__':` since they won't be executed anyway.

I suggest you extract `def collecting_cars` and `def url_read` from under `if __name__ == '__main__':`

File: main.py [hunk 8]
```diff
@@ -1,16 +1,68 @@
...
+        try:
+            with open(filename, 'r', encoding='utf-8') as file:
+                data = json.load(file)
+            return data
+        except FileNotFoundError:
+            print(f"File '{filename}' not found.")
+            return {}
+        except Exception as e:
+            print(f"An error occurred: {str(e)}")
+            return {}
```
Comment (alexpilk) [inline main.py:19]: It's common to add this kind of error handling but it doesn't provide any benefits.

1. If the filename is wrong and the file's not found - you will get an empty result and will have to guess what happened. Instead you would be better off had the script failed with `FileNotFoundError` because you'd know what's up right away.

2. By catching the general `Exception` and printing the error all you're doing is hiding the traceback making debugging harder. It's usually a bad practice to catch general `Exception`s because you're silencing errors you cannot handle anyway. The most common case when it makes sense is if you want to send a notification (say via slack or email) first and then still throw the exception.

3. By adding this try-except you're making the code more complex and also end up with two more edge cases to cover with tests.

File: main.py [hunk 9]
```diff
@@ -1,16 +1,68 @@
...
+                country, cars = list(source_data.items())[0]
+                print('-' * 30)
+                print(f"Country: {country}:")
+                print('-' * 30)
+                for car in cars:
+                    print(f'ID: {car.id}')
+                    print(f'Title: {car.title}')
+                    print(f'Link: {car.link}')
+                    print(f'Price: {car.price}')
+                    print('-' * 30)'''
```
Comment (alexpilk) [inline main.py:60]: Leaving commented out code is bad practice. Instead, you could:

1. Add a feature flag that would either be an argument (if you were to wrap this into a function) or an environment to enable/disable this block of code
2. Use loggers instead of `print` statements. Then you could use the `DEBUG` logging level for this and it would be printed or not based on your logging config

File: main.py [hunk 10]
```diff
@@ -1,16 +1,68 @@
...
+
+            car_data = {
+                car_name: [{"PL": list(pl_data)}, {"PT": list(pt_data)}]
+            }
+
+            all_cars_data.append(car_data)
+
+        return all_cars_data
+
+    # Read car names and URLs from the file using url_read()
```
Comment (alexpilk) [inline main.py:40]: I would delete ALL comments in ALL files because the code is self-explanatory (which is good).

Usually comments only make sense if the code is doing something confusing and there should be a good excuse for it.

File: data_parser.py [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+        numeric_price = float(cleaned_price)
+
+        # Check if the price text contains EUR symbol
+        if '€' in price_text:
+            # If EUR is present, return the numeric value
+            return round(numeric_price)
+        elif 'zł' in price_text:
+            # If PLN is present, convert to EUR
+            price_in_eur = numeric_price / eur_to_pln_rate
+            return round(price_in_eur)
```
Comment (alexpilk) [inline data_parser.py:23]: If the goal of the app is to approximate price differences then you can get away with this, otherwise you'd have to use `Decimal` to avoid rounding errors.

File: analysis.py [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+            coefficients.append({car_name: percentage_difference})
+
+    return coefficients
+
+
+def get_exchange_rate():
+    try:
+        # Define the base URL for the exchange rate API
+        base_url = 'http://api.exchangeratesapi.io/v1/latest?access_key=563a0d673a2f0b71544fb07b856c7686&format=1&symbols=PLN'
```
Comment (alexpilk) [inline analysis.py:53]: You don't want to hardcode API keys. If the app is a script - ask users to supply it as input data.

If it's a web app - the key should be in the environment variables & not committed into the repo.

File: analysis.py [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+        median_price_pt = median(prices_pt)
+
+        # Calculate the coefficient (price ratio) using median prices
+        coefficient = median_price_pt / median_price_pl
+
+        return coefficient
+
+    except Exception as e:
+        print(f'An error occurred: {str(e)}')
+        return None"""
```
Comment (alexpilk) [inline analysis.py:24]:
```suggestion
```

File: analysis.py [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+        prices_pt = [car.price for car in pt_data if car.price != 'N/A']
+
+        if prices_pl and prices_pt:
+            median_price_pl = median(prices_pl)
+            median_price_pt = median(prices_pt)
+
+            price_difference = median_price_pt - median_price_pl
+            percentage_difference = round((price_difference / median_price_pl) * 100, 1)
+
+            coefficients.append({car_name: percentage_difference})
```
Comment (alexpilk) [inline analysis.py:45]: Code duplication, a function should be extracted

File: analysis.py [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+            return pln_to_eur_rate
+
+        else:
+            print(f'Failed to fetch exchange rate. Status code: {response.status_code}')
+
+    except Exception as e:
+        print(f'An error occurred: {str(e)}')
+
+    print(f'[WARN] No rate fetched, using the default')
+    return 4.55
```
Comment (alexpilk) [inline analysis.py:74]: Magic constant, this should be in something like `configuration.py` and named `DEFAULT_EXCHANGE_RATE`

File: main.py [hunk 16]
```diff
@@ -1,16 +1,68 @@
...
+            return {}
+        except Exception as e:
+            print(f"An error occurred: {str(e)}")
+            return {}
 
 
-# Press the green button in the gutter to run the script.
-if __name__ == '__main__':
-    print_hi('PyCharm')
+    def collecting_cars(car_urls):
```
Comment (alexpilk) [inline main.py:22]: Naming:
```suggestion
    def collect_cars(car_urls):
```

File: main.py [hunk 17]
```diff
@@ -1,16 +1,68 @@
...
+
+        for car_name, urls in car_urls.items():
+            pl_url = urls.get("PL", "")
+            pt_url = urls.get("PT", "")
+
+            pl_data = fetch_olx_car_data(pl_url)
+            pt_data = fetch_olx_car_data(pt_url)
+
+            car_data = {
+                car_name: [{"PL": list(pl_data)}, {"PT": list(pt_data)}]
```
Comment (alexpilk) [inline main.py:33]: Why not
```suggestion
                car_name: {"PL": pl_data, "PT": pt_data}
```

**Review by alexpilk — CHANGES_REQUESTED** [review]
